photos/db.go

```diff
@@ -0,0 +1,478 @@
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+	_ "github.com/lib/pq"
+)
+
+const (
+	lgSchema = `
+CREATE DATABASE IF NOT EXISTS photos;
+
+CREATE TABLE IF NOT EXISTS photos.users (
+  id           INT,
+  photoCount   INT,
+  commentCount INT,
+  name         STRING,
+  address      STRING,
+
+  PRIMARY KEY (id)
+);
+
+CREATE TABLE IF NOT EXISTS photos.photos (
+  id           BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
+  userID       INT,
+  commentCount INT,
+  caption      STRING,
+  latitude     FLOAT,
+  longitude    FLOAT,
+  timestamp    TIMESTAMP,
+
+  PRIMARY KEY (id)
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS photosByUserID ON photos.photos (userID, id);
+
+CREATE TABLE IF NOT EXISTS photos.comments (
+  photoID   BYTES,
+  commentID BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
+  userID    INT,
+  message   STRING,
+  timestamp TIMESTAMP,
+
+  PRIMARY KEY (photoID, commentID)
+);
+`
+)
+
+// pgURL returns a Postgresql URL string for the specified context.
+func pgURL(ctx Context) string {
+	certUser := security.RootUser
+	options := url.Values{}
+	options.Add("sslmode", "verify-full")
+	options.Add("sslcert", security.ClientCertPath(ctx.CertsDir, certUser))
+	options.Add("sslkey", security.ClientKeyPath(ctx.CertsDir, certUser))
+	options.Add("sslrootcert", security.CACertPath(ctx.CertsDir))
+	url := url.URL{
+		Scheme:   "postgres",
+		User:     url.User(certUser),
+		Host:     ctx.DBAddr,
+		RawQuery: options.Encode(),
+	}
+	return url.String()
+}
+
+// openDB opens the database connection according to the context.
+func openDB(ctx Context) (*sql.DB, error) {
+	return sql.Open("postgres", pgURL(ctx))
+}
+
+// initSchema creates the database schema if it doesn't exist.
+func initSchema(db *sql.DB) error {
+	_, err := db.Exec(lgSchema)
+	return err
+}
+
+// dropDatabase drops the database.
+func dropDatabase(db *sql.DB) error {
+	_, err := db.Exec("DROP DATABASE photos;")
+	return err
+}
+
+// execute runs a closure within a transaction. On failure,
+// the transaction is aborted and rolled back; on success,
+// the transaction is committed. If the error string matches
+// CockroachDB's restart directive, the same transaction is
+// re-used to retry executing the closure.
+//
+// NOTE: the supplied exec closure should not have external
+//       side effects beyond changes to the database.
+func executeTxn(db *sql.DB, exec func(*sql.Tx) error) error {
+	tx, err := ctx.DB.Begin()
+	if err != nil {
+		return err
+	}
+	for {
+		if err := exec(tx); err != nil {
+			// TODO(spencer): check error return here and if retry txn is
+			// warranted, continue.
+			/*
+				if err.Error().Matches("retry cockroachdb txn") {
+					if _, err = tx.Exec(`RETRY TRANSACTION;`); err == nil {
+						continue
+					}
+				}
+			*/
+			_ = tx.Rollback()
+			return err
+		}
+		return tx.Commit()
+	}
+}
+
+const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+func randStringBytes(n int) string {
```

Comment:
Not sure why you have `Bytes` in the name here: `s/randStringBytes/randString/g`?